src/v/cloud_storage/remote_partition.cc

```diff
@@ -483,55 +369,43 @@ class partition_record_batch_reader_impl final
             co_return;
         }
         _view_cursor = std::move(cur.value());
-        co_await _view_cursor->with_manifest(
+        auto lookup_result = co_await _view_cursor->with_manifest(
           [this, config](const partition_manifest& manifest) {
-              initialize_reader_state(manifest, config);
+              return _partition->lookup_segment(manifest, config);
           });
-        co_return;
-    }
 
-    // Initialize object using remote_partition as a source
-    void initialize_reader_state(
-      const partition_manifest& manifest,
-      const storage::log_reader_config& config) {
-        vlog(
-          _ctxlog.debug,
-          "partition_record_batch_reader_impl initialize reader state");
-        auto [reader, next_offset] = find_cached_reader(manifest, config);
-        if (reader) {
-            _reader = std::move(reader);
-            _next_segment_base_offset = next_offset;
-            return;
-        }
-        vlog(
-          _ctxlog.debug,
-          "partition_record_batch_reader_impl initialize reader state - "
-          "segment not "
-          "found");
-        _reader = {};
-        _next_segment_base_offset = {};
-    }
+        if (!lookup_result.has_value()) {
+            vlog(
+              _ctxlog.debug,
+              "partition_record_batch_reader_impl initialize reader state - "
+              "segment not found: {}",
+              lookup_result->meta);
 
-    remote_partition::borrow_result_t find_cached_reader(
-      const partition_manifest& manifest,
-      const storage::log_reader_config& config) {
-        if (!_partition || _partition->_manifest_view->stm_manifest().empty()) {
-            return {};
+            co_return;
         }
-        auto res = _partition->borrow_next_segment_reader(manifest, config);
-        if (res.reader) {
-            // Here we know the exact type of the reader_state because of
-            // the invariant of the borrow_reader
+
+        _reader = co_await _partition->borrow_segment_reader(
```

Comment:
at this point `lookup_result->meta` could be invalidated by compacted reupload or adjacent segment merger

Reply:
Thinking about it some more, this can happen with the current code too. Requests are async, so by the time we actually send the request the segment may have been removed and we expect the client to retry.

Reply:
I'll try to look into this more closely when I'm done with time queries. I think we need a bit more general mechanism for this and https://github.com/redpanda-data/redpanda/issues/12105 
Sone coordination when we're doing GC/retention is needed.

Reply:
@Lazin wanna put something in the [backlog](https://github.com/orgs/redpanda-data/projects/112/views/1) such that we don't forget about it?